src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java

```diff
@@ -0,0 +1,524 @@
+     * of {@link RandomGenerator} that utilizes the {@code name} algorithm.
+     *
+     * @param name  Name of random number algorithm to use
+     * @param category Sub-interface of {@link RandomGenerator} to type check
+     * @param <T> Sub-interface of {@link RandomGenerator} to produce
+     *
+     * @return Factory of {@link RandomGenerator}
+     *
+     * @throws IllegalArgumentException when either the name or category is null
+     */
+    static <T extends RandomGenerator> RandomGeneratorFactory<T> factoryOf(String name, Class<? extends RandomGenerator> category)
+            throws IllegalArgumentException {
+        Provider<? extends RandomGenerator> uncheckedProvider = findProvider(name, category);
+        return new RandomGeneratorFactory<>(uncheckedProvider);
+    }
+
+    /**
+     * Fetch the required constructors for class of random number algorithm.
+     *
+     * @param randomGeneratorClass class of random number algorithm (provider)
+     */
+    @SuppressWarnings("unchecked")
+    private void getConstructors(Class<? extends RandomGenerator> randomGeneratorClass) {
+        if (ctor == null) {
+            synchronized (provider) {
+                if (ctor == null) {
+                    PrivilegedExceptionAction<Constructor<?>[]> ctorAction = randomGeneratorClass::getConstructors;
+                    try {
+                        Constructor<?>[] ctors = AccessController.doPrivileged(ctorAction);
+
+                        Constructor<T> tmpCtor = null;
+                        Constructor<T> tmpCtorLong = null;
+                        Constructor<T> tmpCtorBytes = null;
+                        for (Constructor<?> ctorGeneric : ctors) {
+                            Constructor<T> ctorSpecific = (Constructor<T>) ctorGeneric;
+                            final Class<?>[] parameterTypes = ctorSpecific.getParameterTypes();
+
+                            if (parameterTypes.length == 0) {
+                                tmpCtor = ctorSpecific;
+                            } else if (parameterTypes.length == 1) {
+                                Class<?> argType = parameterTypes[0];
+
+                                if (argType == long.class) {
+                                    tmpCtorLong = ctorSpecific;
+                                } else if (argType == byte[].class) {
+                                    tmpCtorBytes = ctorSpecific;
+                                }
+                            }
+                        }
+
+                        if (tmpCtor == null) {
+                            throw new IllegalStateException("Random algorithm " + name() + " is missing a default constructor");
+                        }
+
+                        // Store specialized constructors first, guarded by ctor
+                        ctorBytes = tmpCtorBytes;
+                        ctorLong = tmpCtorLong;
+                        ctor = tmpCtor;
```

Comment:
This one is a volatile write, so the idea is that ctorBytes and ctorLong does not need to be volatile too, which i think is not true if there is a code somewhere that uses ctorBytes or ctorLong without reading ctor.
This code is too smart for me, i'm pretty sure it is wrong too on non TSO cpu.

Reply:
The 2 uses call ensureConstructors, which calls this method, which reads ctor.
The memory model guarantees this to be safe, even on non tso hardware.
https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#pitfall-avoiding